### PR TITLE
Enable GH-Pages Documentation

### PR DESCRIPTION
**Pull request by t8y8**
@RussTheAerialist @benlower @LGraber ,

@jdomingu and I spent some time getting a docs pipeline working for the document api. This can extend to our other repos.

We approached it with the following requirements:
- Documentation should be generated from Markdown
- It should support fenced code blocks and tables (GFM)
- The build should be locally testable
- The deploy should be automatic on push to master
- The tooling should be fairly simple to install/run

These items were considered nice to have but not a hard requirement:
- Auto-Generated API documentation

My prototype made use of mkdocs because it's python, simple, and integrated nicely with travis.
I have a preference, can you tell :)

Mkdocs had a few cons:
- Fine-Grained control over layout (custom layouts aren't supported until 0.16 which is unreleased)
- Deploy to GH-Pages required a hand-written script in the repo and run from travis (But I got it working!!!)

The second approach, presented here, was to move to using the built in auto-generation-and-publish-to-gh-pages via Jekyll.

This has a few advantages:
- No deploy script
- Built in to Github
- We can reuse the template(s) we're already using for the WDC

And a few disadvantages:
- There are no build failures that we can check in travis, they will publish broken docs
- The toolchain is more annoying (Gem, Bundle, Etc) than 'pip install' and you're off

To test docs changes locally:
`bundle exec jekyll serve`

Preview at: https://t8y8.github.io/document-api-python/

Once merged, we can enable GH-Pages and all pushes to master will get updates to the docs automatically.

I'm super excited about this, because I think accessible markdown documentation is important for our users to understand the library, and also as a good place to start contributing. Many first-timers start with docs!

### CODE REVIEW

File: docs/css/extra.css [hunk 1]
```diff
@@ -0,0 +1,14 @@
+pre code {
+  white-space: pre;
+  word-wrap: normal;
+  display: block;
+  padding: 12px;
+  font-size: 14px;
+}
+
+code {
```
Comment (t8y8) [inline docs/css/extra.css]: Here's an example of how we can override some stuff in the themes.

I thought the code blocks were too small

File: docs/contributing.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+For a feature request, please try to describe the scenario you are trying to accomplish that requires the feature.  This will help us understand
+the limitations that you are running into, and provide us with a use case to know if we've satisfied your request. Psuedo code that describes how you would like things to work is also very helpful.
+
+## Fixes, Implementations, and Documentation
+
+For all other things, please submit a PR that includes the fix, documentation, or new code that you are trying to contribute.  More information on
+creating a PR can be found in the [github documentation](https://help.github.com/articles/creating-a-pull-request/)
+
+If the feature is complex or has multiple solutions that could be equally appropriate approaches, it would be helpful to file an issue to discuss the
+design trade-offs of each solution before implementing, to allow us to collectively arrive at the best solution, which most likely exists in the middle somewhere.
```
Comment (RussTheAerialist) [inline docs/contributing.md]: Couldn't this be a symlink to the upper-level contributing.md (I think we have one)

File: docs/user-guide.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+# Concepts
+
+Imagine a world without kittens. What a world that would be.
```
Comment (RussTheAerialist) [inline docs/user-guide.md]: okay seriously, last night we were playing the expansion to Betrayal at the House on the Hill, and the event was literally, you have to soothe kittens while preparing cat food for the large cat.

File: docs/contributing.md (hunk 2 above)
Comment (t8y8) [inline docs/contributing.md]: Yes! Apparently I can!

File: docs/user-guide.md (hunk 3 above)
Comment (RussTheAerialist) [inline docs/user-guide.md]: Literally my favorite game of all times.

**Comment by RussTheAerialist** [comment]
Games aside, I like this approach.  I wish there was a way to generate docs from the docstrings, but this is probably good enough.

**Comment by t8y8** [comment]
There are a few experimental plugins for mkdocs that we may be able to leverage as things mature:

- https://github.com/mkdocs/mkdocs/issues/206
- https://github.com/erichulser/mkdocs_autodoc/blob/master/README.md

It looks like the mkdocs author is mostly concerned with having a clean simple feature set (mkdocs), and has a few ideas for a plugin api that could handle the rest. THere's active discussion on their tracker.

I think it's ok, with the size of our current projects, to have manual docs -- and we can integrate docstring-generated items in the future

File: docs/css/github-highlight.css [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+  font-weight: bold;
+}
+.highlight .gu {
+  color: #aaaaaa;
+}
+.highlight .gt {
+  color: #aa0000;
+}
+.highlight .kc {
+  color: #000000;
```
Comment (LGraber) [inline docs/css/github-highlight.css]: I think this should be #000001

Comment (t8y8) [inline docs/css/github-highlight.css]: @jdomingu ?

Comment (LGraber) [inline docs/css/github-highlight.css]: Hahahahahaha

Comment (jdomingu) [inline docs/css/github-highlight.css]: @Lgraber ? Not sure I understand your comments.

On another note, I agree with @t8y8 's assessment. Jekyll is a bit heavier weight, but in general much more flexible. I think it's worth it in the long run especially once we get close to outgrowing the gh-pages pipeline. For now, the automated publishing that we get by using jekyll makes up for many many other things. 

Also, in my opinion, if the expectation is that users are just editing markdown, it's pretty hard to break the build. Yes, it's possible to commit something with odd formatting, but that wouldn't be caught by a build test anyway. Unless we implement an HTML parser and link validator (whether we go with mkdocs or jekyll really, either one) I don't think that we get a substantial benefit from doc testing. 

TLDR I recommend we go with jekyll for the autopublishing and for consistency with the WDC. +Testing is hard for docs regardless of the framework.

Comment (t8y8) [inline docs/css/github-highlight.css]: Hahaha, my best guess was that you were a SKC fan -- but we all know they lost so...

Comment (benlower) [inline docs/css/github-highlight.css]: 🚀 :shipit:
